Support cancelable SPDY executor stream

arkbriar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Sometimes I find that even simple command like "echo something" will hang there forever when there're lots of SPDY executor streams running. In the last case, I've examined all the call stacks with debug tool and still haven't figured out why. It seems all of the streams stuck on waiting for next frame and the TCP connections were still alive then. But these streams had already been existed for over 10 minutes.

Anyway, I think it will be reasonable if some bad streams can be canceled to avoid resource leaks. So I just enhance the `remotecommand.Executor` interface with a method accepts a context to support cancelable SPDY executor stream.

Also test have been modified to ensure it works.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/client-go/issues/884

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add a method `StreamWithContext` to remotecommand.Executor to support cancelable SPDY executor stream.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

arkbriar commented:
/assign @deads2k

fedebongio commented:
/triage accepted

arkbriar commented:
/assign @liggitt

liggitt commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -48,6 +49,9 @@ type Executor interface {
 	// is set, the stderr stream is not used (raw TTY manages stdout and stderr over the
 	// stdout stream).
```
should this be marked deprecated in favor of StreamContext?

arkbriar replied:
Yeah. I think so too. Will be marked.

arkbriar replied:
@liggitt I've mark it as deprecated in a new commit. Please review again when you have time. Thanks.

aojea commented:
should the context plumbed down to the spdy request too?
https://github.com/kubernetes/client-go/blob/ac207faedfb64acd5b99a2fb309b7044918b4dda/transport/spdy/spdy.go#L71

arkbriar commented:
> should the context plumbed down to the spdy request too?
> https://github.com/kubernetes/client-go/blob/ac207faedfb64acd5b99a2fb309b7044918b4dda/transport/spdy/spdy.go#L71

@aojea For stream executor interface, it's enough to have a context here. 
But yes, for the dialer interface, it still needs a context to let timeout possible.
I'll fix it in this PR. 😁

aojea commented:
> > should the context plumbed down to the spdy request too?
> > https://github.com/kubernetes/client-go/blob/ac207faedfb64acd5b99a2fb309b7044918b4dda/transport/spdy/spdy.go#L71
> 
> @aojea For stream executor interface, it's enough to have a context here.
> But yes, for the dialer interface, it still needs a context to let timeout possible.
> I'll fix it in this PR.

I was trying to understand if this could also fix https://github.com/kubernetes/kubectl/issues/686 , but I think I was wrong, and the spdy and Dialer context are not really needed here as you correctly pointed out.

Feel free to discard my comment since the Dialer seems to be used also in another places and changing the signature can break some external users, and this will unnecessary complicate this, sorry :sweat_smile: 
```
staging/src/k8s.io/client-go/tools/portforward/portforward.go:160:func NewOnAddresses(dialer httpstream.Dialer, addresses []string, ports []string, stopChan <-chan struct{}, readyChan chan struct{}, out, errOut io.Writer) (*PortForwarder, error) {
test/e2e/storage/drivers/proxy/portproxy.go:197:func dial(ctx context.Context, prefix string, dialer httpstream.Dialer, port int) (s *stream, finalErr error) {
```

/ok-to-test

arkbriar commented:
> > > should the context plumbed down to the spdy request too?
> > > https://github.com/kubernetes/client-go/blob/ac207faedfb64acd5b99a2fb309b7044918b4dda/transport/spdy/spdy.go#L71
> > 
> > 
> > @aojea For stream executor interface, it's enough to have a context here.
> > But yes, for the dialer interface, it still needs a context to let timeout possible.
> > I'll fix it in this PR.
> 
> I was trying to understand if this could also fix [kubernetes/kubectl#686](https://github.com/kubernetes/kubectl/issues/686) , but I think I was wrong, and the spdy and Dialer context are not really needed here as you correctly pointed out.
> 
> Feel free to discard my comment since the Dialer seems to be used also in another places and changing the signature can break some external users, and this will unnecessary complicate this, sorry 😅
> 
> ```
> staging/src/k8s.io/client-go/tools/portforward/portforward.go:160:func NewOnAddresses(dialer httpstream.Dialer, addresses []string, ports []string, stopChan <-chan struct{}, readyChan chan struct{}, out, errOut io.Writer) (*PortForwarder, error) {
> test/e2e/storage/drivers/proxy/portproxy.go:197:func dial(ctx context.Context, prefix string, dialer httpstream.Dialer, port int) (s *stream, finalErr error) {
> ```
> 
> /ok-to-test

@aojea Yeah I think [kubernetes/kubectl#686](https://github.com/kubernetes/kubectl/issues/686) is probably not-related. Definitely I haven't considered the external users 😅.

aojea commented:
> @aojea Yeah I think [kubernetes/kubectl#686](https://github.com/kubernetes/kubectl/issues/686) is probably not-related. Definitely I haven't considered the external users .

and is breaking the tests because of that :)
Discard the dialer, it can be added later if needed

arkbriar commented:
/test pull-kubernetes-integration

arkbriar commented:
/test pull-kubernetes-integration

Emmm...it says 62/62 passed but still failed.

arkbriar commented:
> @arkbriar: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-integration	[61086ea](https://github.com/kubernetes/kubernetes/commit/61086eab06bb1323798a5cbd20018c75b1b07ce0)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103177/pull-kubernetes-integration/1419659846940102656)	`/test pull-kubernetes-integration`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=103177). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Aarkbriar). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

Rebased on master to see if it will pass.

liggitt commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -106,7 +113,13 @@ func NewSPDYExecutorForProtocols(transport http.RoundTripper, upgrader spdy.Upgr
+	req, err := http.NewRequestWithContext(ctx, e.method, e.url.String(), nil)
 	if err != nil {
 		return fmt.Errorf("error creating request: %v", err)
 	}
```
are there still parts inside spdy.Negotiate that can get stuck without honoring the context (specifically inside the `upgrader.NewConnection(resp)` call... will `ioutil.ReadAll(resp.Body)` unblock if the request context is closed)?

arkbriar replied:
Yes, good question. Actually I've found another case that the context isn't being honored exactly in spdy.Negotiate, but not in upgrader.NewConnection. I've dived in to the stdlib's  http client code and found that resp.Body will always be wrapped with timeout in the following code:

```go
func send(ireq *Request, rt RoundTripper, deadline time.Time) (resp *Response, didTimeout func() bool, err error) {
        ....
	if !deadline.IsZero() {
		resp.Body = &cancelTimerBody{
			stop:          stopTimer,
			rc:            resp.Body,
			reqDidTimeout: didTimeout,
		}
	}
        ...
}
```

The problem lays in the customized http.Transport returned by apimachinery/pkg/util/httpstream/spdy.NewRoundTripperWithConfig which is used in our SPDY executor. It doesn't recognize the context in request and performs lots of the new http requests & response reading without deadline. A simple way to reproduce is performing exec (timeout 1s) frequently (~ 2s) on a single pod and then shutdown it's host. With high probability you will find a stuck exec which typically timeouts after 30 seconds.

![image](https://user-images.githubusercontent.com/8898435/128803389-f2c33edc-a893-4656-9624-c32d4155315f.png)

I have a fix just for my project but isn't ready for submitting a PR to upstream. (I've changed some of the public APIs' signatures). As a reference, you can find it [here](https://github.com/arkbriar/apimachinery/compare/v0.21.2...v0.21.2-enc2).

Anyway, I think it's another issue (transport not respecting context) outside this PR (provide a way to support cancelable).

arkbriar replied:
cc. @aojea 😉

arkbriar replied:
@liggitt Hi dude, any updates?

liggitt replied:
Sorry for the delay. I can buy that this is an improvement even if there are still parts of the operation that can't be interrupted.

Since this is a library that can be used in a lot of different ways, to be sure this doesn't introduce any regressions, asked a couple questions about behavior around panic and hanging goroutines (would like to see those behaviors exercised with unit tests, not just verified manually once)

liggitt commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -138,5 +151,16 @@ func (e *streamExecutor) Stream(options StreamOptions) error {
+	errorChan := make(chan error, 1)
+	go func() {
+		defer close(errorChan)
+		errorChan <- streamer.stream(conn)
```
does canceling the context close the conn/stream operation or does it leave this goroutine hung?

liggitt replied:
can streamer.stream panic? if so, spawning a goroutine here means the process will exit if that happens, as opposed to propagating to any panic recovery code put in place by the `Stream` caller

arkbriar replied:
> does canceling the context close the conn/stream operation or does it leave this goroutine hung?

Yes, the stream closes after the connection closed (in the deferred statement above).

arkbriar replied:
> can streamer.stream panic? if so, spawning a goroutine here means the process will exit if that happens, as opposed to propagating to any panic recovery code put in place by the `Stream` caller

No I don't think so. But still we could add a panic handler.

arkbriar replied:
@liggitt It's not possible to propagate the panic to caller since the `stream` goroutine may be still running when the method returns at `return ctx.Err()`. By the way, they did use `runtime.HandleCrash` in goroutines such what in `streamProtocolV2.copyStdout` as well, which is started by `stream`.

```go
func (p *streamProtocolV2) copyStdout(wg *sync.WaitGroup) {
	if p.Stdout == nil {
		return
	}

	wg.Add(1)
	go func() {
		defer runtime.HandleCrash()
		defer wg.Done()
		// make sure, packet in queue can be consumed.
		// block in queue may lead to deadlock in conn.server
		// issue: https://github.com/kubernetes/kubernetes/issues/96339
		defer io.Copy(ioutil.Discard, p.remoteStdout)

		if _, err := io.Copy(p.Stdout, p.remoteStdout); err != nil {
			runtime.HandleError(err)
		}
	}()
}
```

So if this panics, the process exits as well. AFAIK, there's no unexpected panic unless I've missed a bug (and it hasn't been triggered so far). So instead of being defensive, I think it's good to let it just panic and quit so that we can locate the bug immediately and have a quick fix. Logging is the best that we can do.

liggitt replied:
> @liggitt It's not possible to propagate the panic to caller since the stream goroutine may be still running when the method returns at return ctx.Err().

it *is* possible, like this:

```diff
diff --git a/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go b/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go
index 4e22a970c35..da6a267b83d 100644
--- a/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go
+++ b/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go
@@ -22,6 +22,7 @@ import (
 	"io"
 	"net/http"
 	"net/url"
+	"sync"
 
 	"k8s.io/klog/v2"
 
@@ -155,6 +156,15 @@ func (e *streamExecutor) newConnectionAndStream(ctx context.Context, options Str
 // StreamWithContext opens a protocol streamer to the server and streams until a client closes
 // the connection or the server disconnects or the context is done.
 func (e *streamExecutor) StreamWithContext(ctx context.Context, options StreamOptions) error {
+	var caughtPanicLock sync.Mutex
+	var caughtPanic error
+	defer func() {
+		caughtPanicLock.Lock()
+		defer caughtPanicLock.Unlock()
+		if caughtPanic != nil {
+			panic(caughtPanic)
+		}
+	}()
 	conn, streamer, err := e.newConnectionAndStream(ctx, options)
 	if err != nil {
 		return err
@@ -163,6 +173,21 @@ func (e *streamExecutor) StreamWithContext(ctx context.Context, options StreamOp
 
 	errorChan := make(chan error, 1)
 	go func() {
+		defer func() {
+			err := recover()
+			if err == nil {
+				return
+			}
+			// Same as stdlib http server code. Manually allocate stack
+			// trace buffer size to prevent excessively large logs
+			const size = 64 << 10
+			buf := make([]byte, size)
+			buf = buf[:runtime.Stack(buf, false)]
+
+			caughtPanicLock.Lock()
+			defer caughtPanicLock.Unlock()
+			caughtPanic = fmt.Errorf("%v\n%s", err, buf)
+		}()
 		defer runtime.HandleCrash()
 		defer close(errorChan)
 		errorChan <- streamer.stream(conn)
```

That ensures existing callers (using context.Background()) have panics propagated as before instead of exiting the process, right?

arkbriar replied:
Yes, I agree that we can propagate panics like this. But it doesn't handle the cases when contexts can be canceled, in which the lock can't block the return of `StreamWithContext` and the panic might happen after that. What about using a channel? Like the following:

```diff
@@ -155,6 +155,13 @@ func (e *streamExecutor) newConnectionAndStream(ctx context.Context, options Str
 // StreamWithContext opens a protocol streamer to the server and streams until a client closes
 // the connection or the server disconnects or the context is done.
 func (e *streamExecutor) StreamWithContext(ctx context.Context, options StreamOptions) error {
+       panicChan := make(chan any, 1)
+       defer func() {
+               if r := <-panicChan; r != nil {
+                       panic(r)
+               }
+       }()
+
        conn, streamer, err := e.newConnectionAndStream(ctx, options)
        if err != nil {
                return err
@@ -163,6 +170,12 @@ func (e *streamExecutor) StreamWithContext(ctx context.Context, options StreamOp

        errorChan := make(chan error, 1)
        go func() {
+               defer func() {
+                       if r := recover(); r != nil {
+                               panicChan <- r
+                       }
+                       close(panicChan)
+               }()
                defer runtime.HandleCrash()
                defer close(errorChan)
                errorChan <- streamer.stream(conn)
```

Do we only care about callers who pass a background context, or do we want to cover all the cases? If it's the first one, we can have a separate branch that runs the `streamer.stream` directly when `Context` is not cancellable, which can be determined with `Context.Done() == nil`. If not and we still want to propagate the panic, then I think blocking the caller with a channel in defer is a feasible solution. But still, I have to say, we can not propagate panics inside the goroutines started by `streamer.stream` unless we modify them in such a way as well.

What do you think?

aojea replied:
>  in which the lock can't block the return of StreamWithContext and the panic might happen after that.

at that time this has already returned, do we want to crash or swallow the panic? it seems that trying to cover everything sounds too much ... and this scenario seems very rare

arkbriar replied:
> > in which the lock can't block the return of StreamWithContext and the panic might happen after that.
> 
> at that time this has already returned, do we want to crash or swallow the panic? it seems that trying to cover everything sounds too much ... and this scenario seems very rare

IMO, I think it's ok not to handle the panic because I believe it won't, as far as I understand. And I agree that we don't have to cover everything, but I prefer to keep the behavior predictable. So my suggestion would be either we use `runtime.HandleCrash` here as others did and do nothing more, or we propagate the panic here using a channel. WDYT?

aojea replied:
I mean, we can propagate the panic, I think that is what Jordan suggested and is ok, but as you said there is still a case where the panic could not be propagated because we already returned, my question is about that case, what are we going to do it, crash the program or ignore it because the context was already cancelled?

aojea replied:
basically I'm asking if jordan's suggestion with the var and mutex is enough ... this is becoming more complex, but Jordan intuition use to be right

liggitt replied:
I agree it's complex, I just don't have visibility to how this library is being used out of tree, and what the increased risk of unhandled panic is, and what the implications for those consumers are :-/

aojea replied:
nah, panic should be handled for sure, if it is used in a handler we can not shutdown the whole server .. does it means that we can swallow the panic after the context has cancelled 🤔 ?

arkbriar replied:
Unless we have something like the structured concurrency in Go, we always increase the risk of unhandled panic when starting a new goroutine. 

I agree that to make sure the program does not exit accidentally, the best choice is to handle the panic thrown by the new goroutine. But I don't think we can swallow the panic. It could confuse the programmer who uses this method:

```go
ctx, cancel := context.WithTimeout(context.Background(), 5 * time.Second)
defer cancel()
// If the panic happens after the timeout, then err = context.ErrDeadlineExceeded and 
// the panic will be swallowed. 
//
// But if the panic happens before the timeout, then the panic is propagated to the caller.
err := exec.StreamWithContext(ctx, opts)
```

From the users' perspective, they will think these are random panics, not they have gotten something wrong. Therefore, I would prefer to propagate the panic in any case. And we can keep the behavior consistent before and after this change. 🤔 

If this makes sense to you, I'm happy to add the codes.

aojea replied:
sgmt, wait for @liggitt ack

aojea replied:
@liggitt I just found we have this pattern https://github.com/kubernetes/kubernetes/pull/103177/files#discussion_r1001353834 in 

https://github.com/kubernetes/kubernetes/blob/4a3bac5eaea81cef0c7b1dcbf47c9eb5c58a30af/staging/src/k8s.io/client-go/tools/cache/reflector.go#L368-L420

WDYT about using the channel instead of the lock?

liggitt replied:
I don't feel strongly, whatever is clearer

arkbriar replied:
I see. So let me propagate the panic with a channel, shall I? @liggitt @aojea

aojea replied:
channel is used in the std library https://github.com/golang/go/blob/81efd7b347dd6d7f12fd49c6eee0274005734c71/src/net/http/server.go#L3356-L3383

and in other projects too https://grep.app/search?q=panicCh&filter[lang][0]=Go

aojea replied:
let's go with channel

arkbriar replied:
Done! Please take a look. @aojea @liggitt 

I followed the pattern from https://github.com/golang/go/blob/81efd7b347dd6d7f12fd49c6eee0274005734c71/src/net/http/server.go#L3356-L3383

arkbriar commented:
> @arkbriar: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-conformance-kind-ipv6-parallel	[3623d4b](https://github.com/kubernetes/kubernetes/commit/3623d4b3ac9dce12af1736914600a25e09154ff4)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103177/pull-kubernetes-conformance-kind-ipv6-parallel/1430380441629626368)	`/test pull-kubernetes-conformance-kind-ipv6-parallel`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=103177). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Aarkbriar). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

Log shows that the build env uses go 1.16.7 but k8s requires go 1.17.

liggitt commented:
/retest

arkbriar commented:
/retest

enj commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -43,11 +46,17 @@ type StreamOptions struct {
-	// non-nil stream to a remote system, and return an error if a problem occurs. If tty
-	// is set, the stderr stream is not used (raw TTY manages stdout and stderr over the
-	// stdout stream).
+	// Deprecated: use StreamWithContext instead to avoid possible resource leaks.
+	// See https://github.com/kubernetes/kubernetes/pull/103177 for details.
 	Stream(options StreamOptions) error
```
Is it worth keeping this method around instead of requiring external callers to update?

If we do want to keep it, should we create an issue to remove the deprecated method in a release or two?

arkbriar replied:
@enj Yes, I think we need an issue to notice the users who relies on this.
I'll create one after this PR is merged.

arkbriar commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -264,3 +300,65 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
+	<-started
+	time.Sleep(100 * time.Millisecond)
+
+	// Close the connection
+	conn.Close()
+
+	select {
+	case <-time.After(100 * time.Millisecond):
+		t.Fatalf("expect stream to be closed after connection is closed.")
```
Thanks and good catch. 
I'll extend it to 1s (and it should be long enough :)

arkbriar commented:
/retest

dims commented:
/assign @enj 

Mo: does this look ready now? looks like the last set of chances are in and the CI is green

piepmatz commented:
Regarding the parts that can get stuck but are not touched in this PR (as mentioned in https://github.com/kubernetes/kubernetes/pull/103177#discussion_r685495910): I opened https://github.com/kubernetes/kubernetes/pull/107605 to tackle them.

@arkbriar, in https://github.com/kubernetes/kubernetes/pull/103177#issuecomment-951455031 you mention that the upgrade to SPDY is also affected. Would be great if you could have a look if my PR covers your finding or if you found something additional.

fedebongio commented:
/assign @caesarxuchao

arkbriar commented:
> Regarding the parts that can get stuck but are not touched in this PR (as mentioned in [#103177 (comment)](https://github.com/kubernetes/kubernetes/pull/103177#discussion_r685495910)): I opened #107605 to tackle them.
> 
> @arkbriar, in [#103177 (comment)](https://github.com/kubernetes/kubernetes/pull/103177#issuecomment-951455031) you mention that the upgrade to SPDY is also affected. Would be great if you could have a look if my PR covers your finding or if you found something additional.

@piepmatz Sorry for the late. I'm really glad to review the PR and I'll look into it ASAP.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -108,13 +122,15 @@ func writeMassiveData(stdStream io.Writer) struct{} { // write to stdin or stdou
 	}{
 		{
-			name: "stdoutBlockTest",
+			timeout: 0,
```
nit: no need to spell out the empty value.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -171,16 +207,16 @@ func newTestHTTPServer(f AttachFunc, options *StreamOptions) *httptest.Server {
 	}
 
-	e := make(chan error)
+	e := make(chan error, 1)
```
Why do we need buffering here?

arkbriar replied:
Because the sender blocks if there's no buffer and receiver at the same time. `select` below would quit early if context is cancelled or timeout so that conditions are all met and stream goroutine leaks.

caesarxuchao replied:
Do you mean that without the buffering, if the `select` picks the `time.After` branch, then the `e <- exec.StreamWithContext(ctx, options)` goroutine will block and leak? Good catch.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -264,3 +300,65 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
+	defer cancelFn()
+
+	uri, _ := url.Parse(server.URL)
+	exec, err := newStreamExecutor(&rest.Config{Host: uri.Host}, "POST", uri)
```
Why not just using the real `NewSPDYExecutor` with a type assertion?

arkbriar replied:
Clever way. I haven't thought of that.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -264,3 +300,65 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
+
+	// Wait until stream goroutine starts.
+	<-started
+	time.Sleep(100 * time.Millisecond)
```
Is the purpose of this sleep to give time to the underlying SPDY connection to establish new streams? I guess without this sleep, the streamer can be stuck at this 30s [Wait](https://github.com/kubernetes/kubernetes/blob/d90342e500a43e5112da04ecefa0c9ae1b1fb6c2/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection.go#L141) when `conn.Close()` is called. And because this 30s Wait is uninterruptible, then the test will fail because the `streamer.stream()` will not be cancelled within 1s.

Relying on the 100ms sleep can be flaky. Instead, can we read/write some contents through the streams to indirectly verify that the streams have been established?

arkbriar replied:
No, the sleep is to avoid connection being closed before the stream, which is a false negative. But yes we can have a method to indicate if the stream starts by reading from the stream.

caesarxuchao replied:
> No, the sleep is to avoid connection being closed before the stream, which is a false negative.

Yeah, that's also possible.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -264,3 +300,65 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
+	}, nil
+}
+
+func TestStreamExitsAfterConnectionIsClosed(t *testing.T) {
```
Can you add a comment on why we need this test? Something like "StreamWithContext closes the conn when the context is done. This test verifies that the close of the conn leads to the termination of the stream.stream() goroutine running in the background".

arkbriar replied:
Thought I think the test name is meaningful enough, it's fine to add such comments.

caesarxuchao replied:
Thanks.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -43,11 +45,17 @@ type StreamOptions struct {
+	// transport any non-nil stream to a remote system, and return an error if a problem
+	// occurs. If tty is set, the stderr stream is not used (raw TTY manages stdout and
+	// stderr over the stdout stream).
+	// The context controls the entire lifetime of stream execution including negotiating
```
The negotiation isn't covered yet, right?

caesarxuchao replied:
Yep.

Can you remove the "negotiation" here and ask the author of #107605 to add it back? We can't guarantee that this PR and #107605 will be merged atomically.

arkbriar replied:
Just removed the description about the "negotiating and streaming". I'll comment on #107605 to suggest the revisal later.

caesarxuchao reviewed (commented):
Agree this is a net improvement even though the context doesn't cover the entire process.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -264,3 +297,75 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
+}
+
+// `Executor.StreamWithContext` starts a goroutine in the background to do the streaming
+// and expects the deferred close of the connection leads to the exit of it on cancellation.
```
How about  "... lead to the exit of the goroutine on cancellation" to be a little more clear?

arkbriar replied:
Edited.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -264,3 +297,75 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
+
+func (w *writerWrap) Write(p []byte) (n int, err error) {
+	if !w.closed {
+		w.written <- true
```
nit: no need to send `true`. Just close `w.written` would unblock.

arkbriar replied:
Thx, just removed.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -264,3 +297,75 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
 }
+
+// A writer wrap which provides a helper method to block until written.
+type writerWrap struct {
```
nit: how about `writeDetector`, which is more self-explanatory?

caesarxuchao reviewed (commented):
Some nits. LGTM.

arkbriar commented:
/test pull-kubernetes-integration

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -131,12 +145,31 @@ func TestSPDYExecutorStream(t *testing.T) {
+			ctx, cancelFn = context.WithTimeout(ctx, test.timeout)
+		}
+
+		err := func(ctx context.Context, cancel context.CancelFunc) error {
+			defer cancelFn()
+			return attach2Server(ctx, server.URL, test.options)
+		}(ctx, cancelFn)
```
Why not just

```suggestion
		err := attach2Server(ctx, server.URL, test.options)
		cancelFn()
```
?

arkbriar replied:
They are the same but i wanna defer the cancellation as package `context` recommends. 

And I have a typo 🙃. The `cancelFn` should be `cancel`.

aojea replied:
have you fixed the typo you mentioned?

arkbriar replied:
🥲 Sorry, I forgot to fix this. Just fixed. Now it's good, I think.

caesarxuchao reviewed (commented):
LGTM

pchy93 commented:
i face the same issue, when the issue can be merged?

aojea commented:
/priority important-soon

Let's try to get this early in the cycle , once master opens for 1.26 

@arkbriar can you please squash the commits?

arkbriar commented:
> /priority important-soon
> 
> Let's try to get this early in the cycle , once master opens for 1.26
> 
> @arkbriar can you please squash the commits?

yeah, for sure!

arkbriar committed:
Support cancelable SPDY executor stream

Mark remotecommand.Executor as deprecated and related modifications.

Handle crash when streamer.stream panics

Add a test to verify if stream is closed after connection being closed

Remove blank line and update waiting time to 1s to avoid test flakes in CI.

Refine the tests of StreamExecutor according to comments.

Remove the comment of context controlling the negotiation progress and misc.

Signed-off-by: arkbriar <arkbriar@gmail.com>

aojea commented:
damn, this slipped again @arkbriar do you still want to try?

/cc @enj

arkbriar commented:
> damn, this slipped again @arkbriar do you still want to try?
> 
> /cc @enj

Hmmm, 🥲. Sure, let me resolve the conflicts first. 
Anyway thanks @aojea for driving this.

arkbriar commented:
Hey @aojea, it's up-to-date now. Shall we try again 😄?

aojea commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:
```diff
@@ -130,12 +144,31 @@ func TestSPDYExecutorStream(t *testing.T) {
 	}
 
 	for _, test := range tests {
 		server := newTestHTTPServer(test.attacher, &test.options)
```
we should use subtests and move the `server.Close()` on L180 here and use a defer

arkbriar committed:
use subtests and defer in TestSPDYExecutorStream

Signed-off-by: arkbriar <arkbriar@gmail.com>

aojea commented:
/lgtm
in addition to https://github.com/kubernetes/kubernetes/pull/103177/commits/42808c8343671e6783ba4c901dcd619bed648c3d#pullrequestreview-865357564

/assign @liggit @enj 

for approval

aojea commented:
/assign @liggitt 
this time the good one

arkbriar committed:
Propagate the panic with a channel

Signed-off-by: arkbriar <arkbriar@gmail.com>

arkbriar commented:
/retest-required

arkbriar commented:
/retest-required

liggitt commented:
/lgtm
/approve

liggitt commented:
thanks for all the iterations on this

arkbriar commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

aojea commented:
/retest

failures are related to https://github.com/kubernetes/kubernetes/pull/113548

/lgtm

Thanks @arkbriar for persisting, sorry for the time it took 😄

reckless-huang commented:
> /复检
> 
> 故障与 #113548 有关
> 
> /lgtm
> 
> 感谢您的坚持，对不起，花了😄时间

greate job
